storm-core/src/jvm/org/apache/storm/zookeeper/Zookeeper.java

```diff
@@ -0,0 +1,355 @@
+        int report = 2000;
+        int limitPort = 65535;
+        if (port != null) {
+            report = port;
+            limitPort = port;
+        }
+        while (true) {
+            try {
+                factory = new NIOServerCnxnFactory();
+                factory.configure(new InetSocketAddress(port), 0);
+                break;
+            } catch (BindException e) {
+                report++;
+                if (report > limitPort) {
+                    throw new RuntimeException("No port is available to launch an inprocess zookeeper");
+                }
+            }
+        }
+        factory.startup(zk);
+        return factory;
+    }
+
+    public static void shutdownInprocessZookeeper(Factory handle) {
+        handle.shutdown();
+    }
+
+    public static NimbusInfo toNimbusInfo(Participant participant) {
+        String id = participant.getId();
+        if (StringUtils.isBlank(id)) {
+            throw new RuntimeException("No nimbus leader participant host found, have you started your nimbus hosts?");
+        }
+        NimbusInfo nimbusInfo = NimbusInfo.parse(id);
+        nimbusInfo.setLeader(participant.isLeader());
+        return nimbusInfo;
+    }
+
+    // Leader latch listener that will be invoked when we either gain or lose leadership
+    public static LeaderLatchListener leaderLatchListenerImpl(Map conf, CuratorFramework zk, LeaderLatch leaderLatch) throws Exception {
+        final String hostName = InetAddress.getLocalHost().getCanonicalHostName();
+        return new LeaderLatchListener() {
+            @Override
+            public void isLeader() {
+                LOG.info(hostName + " gained leadership");
+            }
+
+            @Override
+            public void notLeader() {
+                LOG.info(hostName + " lost leadership.");
+            }
+        };
+    }
+
+    public static ILeaderElector zkLeaderElector(Map conf) throws Exception {
+        List<String> servers = (List<String>) conf.get(Config.STORM_ZOOKEEPER_SERVERS);
+        Object port = conf.get(Config.STORM_ZOOKEEPER_PORT);
+        CuratorFramework zk = mkClient(conf, servers, port, "", conf);
+        String leaderLockPath = conf.get(Config.STORM_ZOOKEEPER_ROOT) + "/leader-lock";
+        String id = NimbusInfo.fromConf(conf).toHostPortString();
+        AtomicReference<LeaderLatch> leaderLatchAtomicReference = new AtomicReference<>(new LeaderLatch(zk, leaderLockPath, id));
+        AtomicReference<LeaderLatchListener> leaderLatchListenerAtomicReference =
+                new AtomicReference<>(leaderLatchListenerImpl(conf, zk, leaderLatchAtomicReference.get()));
+        return new LeaderElectorImp(conf, servers, zk, leaderLockPath, id, leaderLatchAtomicReference, leaderLatchListenerAtomicReference);
+    }
+
+    //To do modify @return once don't need persistentArrayMap
+    public static PersistentArrayMap getDataWithVersion(CuratorFramework zk, String path, boolean watch) {
```

Comment:
This does not have to be a PersistentArrayMap.  It can be any Map, but if you want to change it to a specific class right now, and update all of the references to it in the clojure code that is fine with me too.